.azure-pipelines/release.yml

```diff
@@ -0,0 +1,37 @@
+# Release pipeline to build and deploy Certbot for Windows for GitHub release tags
+trigger:
+  tags:
+    include:
+      - v0.*
+pr: none
+
+stages:
+  - stage: build_n_test
+    jobs:
+      - template: templates/tests-suite.yml
+      - template: templates/installer-tests.yml
+  - stage: release
+    jobs:
+      - job: github
+        pool:
+          vmImage: vs2017-win2016
+        steps:
+          - task: DownloadPipelineArtifact@0
+            inputs:
+              artifactName: WindowsInstaller
+              targetPath: $(Build.ArtifactStagingDirectory)
+          - bash: |
+              CERTBOT_VERSION="${BUILD_SOURCEBRANCH/refs\/tags\/v/}"
```

Comment:
Yes. We could trigger the release pipeline not on the tags of type `v*`, but on the release branches of type `*.x`. This way one could just push a new commit to `0.40.x` for instance, in order to trigger again the release pipeline before tagging the specific commit once it is sure that the installer have been properly been build.

In this case, I need to be smarter to get the actual version of Certbot, but something similar to what is occurring in `setup.py` will do the trick.

What do you think of this?

Reply:
Yes my response was specifically on the concern you express. I was designing my proposition around the fact that this pipeline will not push a release on GitHub anymore, but just expose the artifacts to be published afterward by external mean.

What bother me a little with your solution is breaking the single source of truth principle: this pipeline is triggered by a push to the repository, not anything else. This way it is clear from which commit exactly the artifacts come from. My solution keeps that principle valid, since you can just push another commit to the release branch is the first one was unsuccessful.

When the pipeline has confirmed that an installer can be built, then you can tag the commit and proceed to the next steps of the release.

Reply:
I explained during the 1-1 why I want a dedicated release pipeline, but I write it here for the record. So the main reason is not to have a build of the changelog file, which is more a bonus. The main reason is that the release pipeline does not fulfill the same purpose than the advanced one, even if (for now) it does effectively execute the same actions.

As the release pipeline will provide the artifacts for the release, I want it to be clearly separated from the advanced pipeline in the Azure DevOps UI and so the release process. It is supposed to contains deployable artifacts, while the advanced pipeline will most of the time be the result of nightly tests, or tests on demand for the `test-*` branches. In the future, it would also fit naturally in so-called Azure DevOps release pipeline (the CD part), that would typically be triggered by a specific CI pipeline (here, `release.yml`) and the publication of its pipeline artifacts.

Also the risk of divergence between the pipelines is largely mitigated by the use of job templates, as it is done here.

In term of release workflow, it is true that state of the art on CI/CD is more in favor of define the deploy of new release based on a deployable branch than a deployable tag. Typically a release flow is triggered by a push to the deployable branch (invoking the CD at this time), and a tag is added at the end of the flow to mark that deployable artifacts have been effectively deployed. We should certainly go to that direction in the future.

So, in that aspect @bmw, considering that I remove the `*.x` branch trigger from `advanced.yml`, is it better:
* build `release.yml` on push to `*.x` branches
* build `release.yml` on tag push of type `v*`?

Either way I will apply your proposition and calculate the version of certbot from the sources, not from the tag or branch name that is been built.

Reply:
I see the value in making it abundantly clear where to get the artifacts for the release. I just want to explain my position a little more and hear what you think. My small concern with splitting them is other than any deployment steps, I see no reason for the release pipeline and the advanced pipeline to ever diverge at all and running all tests again in a 2nd pipeline feels a little confusing and a waste of resources. Part of me wonders if it's simpler, especially in the long term, to keep everything in one pipeline and just have the final deployment steps conditional on some attribute like the presence of a git tag.

As for what action on GitHub triggers building the release, I've actually come full circle and think we should just build on tags and we won't push to `*.x` any earlier in the release process for now.

To tell you a little more about our release process at a high level so you can tell me what you think about that idea, currently before building the release, we make sure that all nightly tests have passed on the commit we want to release. This would include building the installer.

Once we've verified that, we make minor changes locally such as updating version numbers, dates, and signatures and then run some simple tests such as our unit tests on the built packages and a couple tests on certbot-auto to make sure everything still works. For example, here is the diffs made locally between the heavily tested commit and the release tag for our last release: https://github.com/certbot/certbot/compare/e3dbd9c...v0.39.0. After the minor additional testing has been done, we push the git tag, push the packages to PyPI, and then update `master` and the `*.x` branches in that order.

Now, in the 4 years of releasing Certbot I don't think we've ever had tests pass on the commit before we made changes and then fail on the tag. This and the fact that updating `*.x` earlier and waiting on the results further complicates the already painful release process (especially for point releases where `*.x` already exists as a protected branch), I'm think it's not necessary, we won't push `*.x` early, and we'll just build on the tag. One very nice thing about this is the history of the release pipeline is extremely clean. It's just our git tags.

Theoretically something could go wrong with building the installer, but I think the chances of it actually happening after it was built before our local changes and our tests have passed locally are extremely low and the complications that building it again earlier make to both the history of the release pipeline and the release process in general aren't worth it.

What do you think?

Reply:
I largely agree with you. Running everything (tests + release logic) on a single tag, and not take care of the actual branch or commits on it will simplify the workflow. In the very improbable situation where the first tag would fail to build the release artifacts (or fail on the integration tests), we can just push a new version of it (like `v0.40.0-1`). By resolving the actual version of Certbot independently from the tag name, the new pipeline will retry to push the `0.40.0` version of Certbot.

On that matter, I think it is preferable to always trigger the pipeline through a tag, and not from the Azure DevOps UI. I think there is some valuable insight to keep track through tags in the GIT repo of failed releases pipelines.

Then I think it is clear that the `advanced`/`release` pipeline should be triggered: 
* on push to `test-*` branches
* on push to `v*` tags
* triggered nightly

Now, about the existence of a dedicated `release` pipeline for push to `v*` tags. To go toward your approach, it is true that in a DevOps approach, it is desirable that the integration branch (here `master`) is always deployable. This means that we should be able to extract from it at anytime a proper and stable version of Certbot. On that extend you do not even need anymore dedicated release branches, and just tag from `master`. This is the approach used in the [GitHub flow](https://gitversion.readthedocs.io/en/latest/git-branching-strategies/githubflow/)

This is an open door to Continuous Deployment, since you can trigger a release anytime it is required, even after each PR merged if you want.

In that perspective, there is no dedicated `release` pipeline. There is only the `advanced` one, that constitutes your oracle to say that `master` branch is deployable.

I have only two concerns about having only the `advanced` pipeline:
* we are not on a CD approach _yet_: it means that a lot of human intervention is required to build a release, and I am a little concern that some wrong artifacts could be mistakenly taken from the `advanced` pipeline
* if we go someday to the so-called Azure `Releases Pipelines`, it may be more cumbersome to define a proper trigger if all CI pipelines are merged for the various situations of on-demand tests, nightly tests and release tags builds.

If these two concerns seems reasonable to you, I will go to the merged of the two pipelines. Take into account that having two pipelines is still very lightweight to maintain thanks to the job templates, and most of the work will be done in them, not in the pipelines themselves.

I would like also, if we drop the scripts about the GitHub release changelog, that it is done manually during a Certbot release because I think there is a lot of values to get the dedicated changelog on the associated GitHub release.

What do you think of this?

Reply:
Good point about creating `v0.40.0-1` tags in the rare event something goes wrong not being that big of a deal.

I would still like the `advanced` pipeline to run on commits to the `*.x` branches. It's currently how we're getting all of our tests to be triggered in preparation for a point release.

I know absolutely nothing about Azure "Release Pipelines", but I definitely agree there's value in making it abundantly clear where to take the release from especially when we're doing things manually.

How about we stick with what you initially had? `release.yml` is triggered on `v*` tags and the triggers for `advanced.yml` are unchanged. Sorry for such a long conversation to make no changes, but this was useful for me.

The one nitpick I'd like to make is can we add a comment to both files saying that if you are adding tests to one, you should probably update the other as well?

I don't have a strong preference on way or another about the changelog script. I think it's fine to keep it without changes if you'd like.

Reply:
No problem about the "circle" made here, and in fact, there is still a modification to do, that is to calculate the actual version of Certbot from the sources, not the tag. Anyway these efforts for automation need to be closely coupled to the human workflow around it, so it is always worth the extra discussions.

You are right about building the releases branches, I forgot the point releases in my logic, as they need dedicated branches extracted from the main deployable branch.

I will add the comments you recommend. I think it is also valuable to only insert jobs in these pipeline as templates, not by direct insertion. It will mitigate the risk of divergence since the actual work will be done in the templates.

About the changelog script, I think I will keep it. Even if it is not ultra useful for now, it will definitely when we will automate more the releases.